### PR TITLE
Add request handling

### PR DESCRIPTION
**Pull request by Draggu**
**Stack**:
- #1686
- #1685
- #1684
- #1683
- #1718
- #1682 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: scarb/src/ops/proc_macro_server/mod.rs [hunk 1]
```diff
@@ -1,13 +1,41 @@
...
+use proc_macro_server_api::RpcResponse;
 
 mod connection;
+mod json_rpc;
 
 pub fn start_proc_macro_server(proc_macros: ProcMacroHost) -> Result<()> {
     let connection = Connection::new();
 
-    //TODO
+    for i in 0..4 {
```
Comment (mkaput) [inline scarb/src/ops/proc_macro_server/mod.rs:13]: use `std::thread:available_parallelism`. do not hardcode threading amounts

File: scarb/src/ops/proc_macro_server/json_rpc.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+use serde::Serialize;
+
+#[derive(Serialize)]
+pub struct ErrResponse {
```
Comment (mkaput) [inline scarb/src/ops/proc_macro_server/json_rpc.rs:4]: this should probably reside in proc-macros-server-api crate. I made a related comment about this in that PR: https://github.com/software-mansion/scarb/pull/1678#discussion_r1816600767

File: scarb/src/ops/proc_macro_server/mod.rs
Comment (mkaput) [inline scarb/src/ops/proc_macro_server/mod.rs:1]: please start adding tests for this system starting with this PR

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/ops/proc_macro_server/mod.rs [hunk 3]
```diff
@@ -1,13 +1,41 @@
...
+        let receiver = connection.receiver.clone();
+        let sender = connection.sender.clone();
+
+        std::thread::Builder::new()
+            .name(format!("proc-macro-server-worker-thread-{i}"))
+            .spawn(move || {
+                for request in receiver {
+                    let response = match request.method.as_str() {
```
Comment (mkaput) [inline scarb/src/ops/proc_macro_server/mod.rs:21]: the indentation is going crazy here, please:
1. extract worker body to a separate function
2. extract request method routing to another one

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/ops/proc_macro_server/mod.rs
Comment (Draggu) [reply scarb/src/ops/proc_macro_server/mod.rs:1]: #1718 Started with next PR (new one) to not bloat diff here

**Review by mkaput — APPROVED** [review]

**Review by piotmag769 — APPROVED** [review]

**Pull request merged**